### PR TITLE
Rooms complete (without Firebase connection)

### PR DESCRIPTION
**Pull request by alexrosen45**
This PR finishes all of the logic for rooms display and interactions. It introduces room creation, adding another user to a room, sending messages, loading messages, and easier manipulation of room displays. It also provides tests at the data access object level. All that's left is writing the Firebase connections in this file alone:
```
package data_access;

import entities.auth.DisplayUser;
import entities.auth.User;
import entities.rooms.Message;
import entities.rooms.Room;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import use_case.rooms.Response;
import use_case.rooms.RoomsDataAccessInterface;

// TODO: You need to split this into multiple DAOs
public class FirebaseRoomsDataAccessObject implements RoomsDataAccessInterface {
  @Override
  public Response<List<Message>> loadMessages(Room room, User user) {
    // TODO: You should get the messages corresponding to room
    DisplayUser dummy_display_user = new DisplayUser("foo", "bar");
    Instant timestamp = Instant.now();
    Message message =
        new Message(timestamp, "This is a test message to load.", dummy_display_user.getUid());

    List<Message> messages = new ArrayList<>();
    messages.add(message);

    // Uncomment this to see what happens when you fail to get messages:
    // Response<List<Messages>> response = new Response<List<Messages>>(null);
    // response.setError("Failed to retrieve messages.");
    // return response;

    return new Response<>(messages);
  }

  @Override
  public Response<String> sendMessage(Room room, User user, String message) {
    // TODO: You should get the messages corresponding to room
    System.out.println(message);

    // Uncomment this to see what happens when you fail to send a message:
    // Response<String> response = new Response<String>(null);
    // response.setError("Failed to send message.");
    // return response;

    return new Response<>("success");
  }

  @Override
  public boolean validateRoomAccess(Room room, User user) {
    // TODO: You should validate that user is a member of room
    // This does not need to be of type Response
    return true;
  }

  @Override
  public Response<String> addUserToRoom(Room room, User user, String email) {
    // TODO: You should add "user" with email "email" to room "room"
    // user being a member of room has already been validated at this point

    // Uncomment this to see what happens when you fail to create room:
    // Response<String> response = new Response<String>(null);
    // response.setError("Failed to add user.");
    // return response;

    return new Response<>("User added successfully!");
  }

  @Override
  public Response<Room> createRoom(User user, String roomToCreateName) {
    // TODO: You should create a room with members {user} and return this new room as a Room entity
    DisplayUser me = new DisplayUser(user.getUid(), user.getName());
    List<DisplayUser> users = new ArrayList<>();
    List<Message> messages = new ArrayList<>();
    users.add(me);

    // Uncomment this to see what happens when you fail to create room:
    // Response<Room> response = new Response<Room>(null);
    // response.setError("Failed to create room.");
    // return response;

    Room room = new Room("1234", roomToCreateName, users, messages);
    return new Response<>(room);
  }
}
```

### CODE REVIEW

File: src/use_case/rooms/Response.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package use_case.rooms;
+
+public class Response<T> {
+  private String error; // Nullable field
+  private T val;
+
+  public Response(T val, String error) {
+    this.val = val;
+    this.error = error;
+  }
```
Comment (uoftm) [inline src/use_case/rooms/Response.java:10]:
```suggestion
  }
  public Response(T val) {
    this.val = val;
    this.error = null;
  }
```

File: src/view/RoomsView.java [hunk 2]
```diff
@@ -94,38 +150,73 @@ public void keyReleased(KeyEvent e) {}
...
+
+    addUserButton.addActionListener(
+        evt -> {
+          if (evt.getSource().equals(addUserButton)) {
+            RoomsState currentState = viewModel.getState();
+            String email = currentState.getEmail();
+            String roomUid = currentState.getRoomUid();
+            if (email != null) {
+              User user = currentState.getUser();
+              for (var room : currentState.getAvailableRooms()) {
```
Comment (uoftm) [inline src/view/RoomsView.java:172]: this logic should probably be in the interactor with a method called getRoomByUid

File: tests/rooms/FirebaseRoomsDataAccessObjectTest.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+import entities.rooms.Message;
+import entities.rooms.Room;
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.List;
+import org.junit.Test;
+import use_case.rooms.Response;
+
+public class FirebaseRoomsDataAccessObjectTest {
```
Comment (uoftm) [inline tests/rooms/FirebaseRoomsDataAccessObjectTest.java:17]: Yay tests!

File: src/view/RoomsView.java [hunk 4]
```diff
@@ -72,6 +77,57 @@ public void keyPressed(KeyEvent e) {}
...
+            // Handle in keyPressed
+          }
+
+          @Override
+          public void keyPressed(KeyEvent e) {
+            RoomsState currentState = viewModel.getState();
+            String currentCreateRoom = currentState.getCreateRoom();
+            String createRoom = currentCreateRoom != null ? currentCreateRoom : "";
+
+            if (e.getKeyCode() == KeyEvent.VK_BACK_SPACE || e.getKeyCode() == KeyEvent.VK_DELETE) {
```
Comment (uoftm) [inline src/view/RoomsView.java:115]: :cry: If we really need this, I'd rather use a document listener on the textinputfield (with a comment indicating that we need to handle backspaces):


```java
createRoomTextField.getDocument().addDocumentListener(
  new DocumentListener() {
                @Override
                public void insertUpdate(DocumentEvent e) {
                    update();
                }

                @Override
                public void removeUpdate(DocumentEvent e) {
                    update();
                }

                @Override
                public void changedUpdate(DocumentEvent e) {
                    update();
                }

                protected void update() {
                    String text = createRoomTextField.getText();

                }
})

File: src/interface_adapter/rooms/RoomsState.java [hunk 5]
```diff
@@ -11,13 +11,20 @@ public class RoomsState {
   private List<Room> availableRooms;
   private List<Message> displayMessages;
   private String sendMessage;
+  private String email; // Other user email
+  private String createRoom; // Room to create
```
Comment (uoftm) [inline src/interface_adapter/rooms/RoomsState.java:15]: The naming of this field confused me for a min, maybe we could change the comment to say "the value of the room creation text field .name" or something?

**Review by uoftm — COMMENTED** [review]
Looking awesome, love the Response class!

File: src/view/RoomsView.java (hunk 2 above)
Comment (alexrosen45) [reply src/view/RoomsView.java:172]: I actually had this written down on paper. Guess I forgot lol.

Comment (alexrosen45) [reply src/view/RoomsView.java:172]: It shouldn't be in the interactor, should be in `RoomsState`.

File: src/interface_adapter/rooms/RoomsState.java (hunk 5 above)
Comment (alexrosen45) [reply src/interface_adapter/rooms/RoomsState.java]: Fixed

File: src/view/RoomsView.java (hunk 4 above)
Comment (alexrosen45) [reply src/view/RoomsView.java:116]: I'll refactor this in a future pr that finalizes `RoomsView` cosmetics.

File: src/use_case/rooms/Response.java (hunk 1 above)
Comment (csc207jc) [reply src/use_case/rooms/Response.java:9]: I think this is a good suggestion @alexrosen45.

Comment (alexrosen45) [reply src/use_case/rooms/Response.java:9]: Yeah, I just implemented this in last commit - forgot to mention.

**Review by csc207jc — APPROVED** [review]
Looks good, I'm ready to implement the Firebase logic for this in the morning!

**Pull request merged**